l2beat/lz-monitoring
# Add changes indexer

Resolves L2B-2877

## What's changed
- Added indexer for changelog
- Pairs of outputs are compared and transformed into granular entries
- Changes indexer relies on discovery indexer
- Contracts can be added/removed - those are represented as milestones to track contract deployments
- Contract's properties are tracked via changelog entries in a separate table
- Each chain configuration now have changelog whitelist - only listed addresses will be taken into account while performing comparison
- There are quite a few tests :trollface: and comments, feel free to question

## Notes
- Logic might be simplified and not split into buckets (`added`, `removed`, `modified`). Also, those buckets are defined from contract standpoint - not the property. Within a `modified` bucket, you won't only see properties that have been modified. It simply means that the given contract is not a newly added one. 

## How to test it?
- Prune database
- Pull latest production dump
- Enable any chain you would like to perform diffing for
- Run application on fresh data to perform migration
- Once indexer runs, take a peak into milestones and changelog tables to see the actual changes that have been gathered based on already present discovery outputs


## Result
![image](https://github.com/l2beat/lz-monitoring/assets/67391475/42a801d5-d774-4553-84b3-d14c6a235fea)
![image](https://github.com/l2beat/lz-monitoring/assets/67391475/308933d2-596b-4cac-ba47-c4daa5b0a3fe)


## Done
- [ ] Indexer test
- [x] Repositories' tests
- [x] Test for milestones
- [x] Cleanup getChangelogChanges - maybe split, maybe not
- [x] Cleanup indexer
- [x] Add test when contract has not values - wut?
- [x] Test outstanding funcs
- [x] Ignore list of changes - optional matches
